File: python/second/combinations.py
class Solution(object):
    # O(n * 2^n) time | O(n) space
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        result = []
        part = []

        def backtrack(s):
            if len(part) == k:
                result.append(part[::])
                return
            
            for i in range(s, n + 1):
                part.append(i)
                backtrack(i + 1)
                part.pop()
        
        backtrack(1)

        return result
    
solution = Solution()
print(solution.combine(4, 2))

File: src/main/java/june_2025/Combinations.java
package june_2025;

import java.util.ArrayList;
import java.util.List;

public class Combinations {

    public static void main(String[] args) {
        int n = 4;
        int k = 2;

        Combinations combinations = new Combinations();
        List<List<Integer>> result = combinations.combine(n, k);
        System.out.println(result);
    }

    // O(n * 2^n) time | O(n) space
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> part = new ArrayList<>();
        backtrack(n, k, 1, part, result);
        return result;
    }

    private void backtrack(int n, int k, int s, List<Integer> part, List<List<Integer>> result) {
        if (part.size() == k) {
            result.add(new ArrayList<>(part));
            return;
        }

        for (int i = s; i <= n; i++) {
            part.add(i);
            backtrack(n, k, i + 1, part, result);
            part.remove(part.size() - 1);
        }
    }


}
